storage/operation/reads.go

```diff
@@ -0,0 +1,207 @@
+type IterationFunc func() (CheckFunc, CreateFunc, HandleFunc)
+
+// IterateKeysInPrefixRange will iterate over all keys with prefixes in the range [startPrefix, endPrefix] (both inclusive)
+func IterateKeysInPrefixRange(startPrefix []byte, endPrefix []byte, check func(key []byte) error) func(storage.Reader) error {
+	return Iterate(startPrefix, endPrefix, func() (CheckFunc, CreateFunc, HandleFunc) {
+		return func(key []byte) (bool, error) {
+			err := check(key)
+			if err != nil {
+				return false, err
+			}
+			return false, nil
+		}, nil, nil
+	}, storage.IteratorOption{IterateKeyOnly: true})
+}
+
+// Iterate will iterate over all keys with prefixes in the given range [startPrefix, endPrefix] (both inclusive)
+func Iterate(startPrefix []byte, endPrefix []byte, iterFunc IterationFunc, opt storage.IteratorOption) func(storage.Reader) error {
+	return func(r storage.Reader) error {
+
+		if len(startPrefix) == 0 {
+			return fmt.Errorf("startPrefix prefix is empty")
+		}
+
+		if len(endPrefix) == 0 {
+			return fmt.Errorf("endPrefix prefix is empty")
+		}
+
+		// Reverse iteration is not supported by pebble
+		if bytes.Compare(startPrefix, endPrefix) > 0 {
+			return fmt.Errorf("startPrefix key must be less than or equal to endPrefix key")
+		}
+
+		it, err := r.NewIter(startPrefix, endPrefix, opt)
+		if err != nil {
+			return fmt.Errorf("can not create iterator: %w", err)
+		}
+		defer it.Close()
+
+		for it.First(); it.Valid(); it.Next() {
+			item := it.IterItem()
+			key := item.Key()
+
+			// initialize processing functions for iteration
+			check, create, handle := iterFunc()
+
+			keyCopy := make([]byte, len(key))
+
+			// The underlying database may re-use and modify the backing memory of the returned key.
+			// Tor safety we proactively make a copy before passing the key to the upper layer.
+			copy(keyCopy, key)
+
+			// check if we should process the item at all
+			shouldReadValue, err := check(keyCopy)
+			if err != nil {
+				return err
+			}
+			if !shouldReadValue { // skip reading value
+				continue
+			}
+
+			err = item.Value(func(val []byte) error {
+
+				// decode into the entity
+				entity := create()
+				err = msgpack.Unmarshal(val, entity)
+				if err != nil {
+					return irrecoverable.NewExceptionf("could not decode entity: %w", err)
+				}
+
+				// process the entity
+				err = handle()
+				if err != nil {
+					return fmt.Errorf("could not handle entity: %w", err)
+				}
+
+				return nil
+			})
+
+			if err != nil {
+				return fmt.Errorf("could not process value: %w", err)
+			}
+		}
+
+		return nil
+	}
+}
+
+// Traverse will iterate over all keys with the given prefix
+func Traverse(prefix []byte, iterFunc IterationFunc, opt storage.IteratorOption) func(storage.Reader) error {
+	return Iterate(prefix, prefix, iterFunc, opt)
+}
+
+// Exists returns true if a key exists in the database.
+// No errors are expected during normal operation.
+func Exists(key []byte, keyExists *bool) func(storage.Reader) error {
```

Comment:
I agree, the API that uses the pointer as argument is a bit confusing. This is the existing badger style and I just followed it.

The reason the badger style has to take pointer as argument is that, the badger read queries has to work with badger transactions, which requires all queries to be curried as functors that takes a single badger transaction obj as argument. 

However, since the new universal database operations is gonna move away from badger transaction, it's possible that all read queries can be made without functors. 

I can still keep this functor style and make another `KeyExists` in the simplified version.

Reply:
In terms of the KeyExists function implementation, there is some detail:

1. I prefer to return `(hasKey bool, exception error)`, instead of just an `error` which use has to compare with `storage.ErrNotFound`
2. Your implementation always call `closer.Close()` regardless there is error or not. It seems the user might get confused about whether Close() call is needed when err != nil. Currently it would return `nil`, which might cause panic if user get confused. So to prevent mistakes like this, I decided to return always return a noopCloser even if err != nil.